---
title: "R Notebook"
output: html_notebook
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(kmerize)
```

## Installation

### This package

```{r eval=FALSE}
install.packages("devtools")
devtools::install_github("c5sire/kmerize")
```


### Supporting tools

This package uses 3rd party tools for basic optimized k-mer creation. Currently, this is the KMC3 set of tools (REF).
The toolkit needs to be installed via the provided function. If no reference to a local file is provided,
the appropriate version for your local operating sytem (OS) will automatically be chosen as shown below.

```{r eval=FALSE}
kmerize::kmr_install_kmc()
```



## What Are K-mers?

K-mers are DNA sequences of length k. The can be generated by moving along a mother- or template DNA in steps of 1. For k = 1 - the shortest - this equals the set of nucleotides A, C, G, and T. Other well-established k-mers include dimers and trimers. Here we are interested in longer k-mers typically in the range from k > 3 to k-mers of lenght < 100 or also occasionally longer.

### What are they good for?

K-mers have found a variety of applications. Particularly interesting is their use as fingerprints at different levels of resolution. In metagenomics, for example, the mixed sample of DNA from a host of organsims is typically annotated with species labels by screening against a database of k-mer signatures.

## Main challenges

A major challenge is to decide on a minimal k-mer size that still is capable to support the envisioned application. The longer the k-mers become the more combinations are possible and so an index in a database will become very big or too big. Along with this, the time to create and consult the database will increase. While this can be remedied to a certain extent with parallel computing, it is still desirable and cheaper, to keep the k-mer size small.

Due to the many still unknown factors in genome biology it is yet not possible to calculate exactly how many distinct k-mers to expect for a certain k and a given species. In general we expect, with increasing k to have more unique k-mers. We also expect, that within a species the unqiue k-mers are limited.

In principle, it is possible to heuristically estimate the total number of k-mers, the distinct, and the unique k-mers by systematically scanning a genome for a certain k range. Experience has shown that it is convenient to only consider kmers of of uneven numbers. 

Here, the example of the virus phix174 (the first sequenced organism) shows the principle following a blogpost here(x). Another wording is, to evaluate the genomic response to changes in k.

### Genomic Response of Phi x 174 

This package contains the full sequende of PhiX174 from NCBI. 

```{r echo=FALSE}
  # Getting a genome
  fa <-  system.file("testdata/phix174.fasta", 
                     package = "kmerize")

  # Setting an output file in a temporary directory since this is for demo
  out_file <- file.path(tempdir(), "phi174")
  
  # here we create a sequence of k in steps of 2
  k <- as.integer(seq(1, 21, 2))
```

These k values are: `r paste(k, sep = ", ")`.

The convenience method for the genomic response returns a summary table. This includes also the theoretical maximum number of k-mers (the k-mer universe) for each value of k.
```{r message=FALSE}
phix_response <- kmr_response(fa, k)
```

The resulting table shows only values up to 13. This is because currently, it stops once the number of unique k-mers equals (exactly more than .9999) the distinct k-mers. This means, at and after this k value all k-mers have become unique.

```{r}
knitr::kable(phix_response)
```
Plotting this table shows the following response curve:

```{r warning=FALSE, fig.width=7}
plot_response(phix_response)
```


