<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<meta name="progressive" content="false" />
<meta name="allow-skip" content="false" />

<title>K-mer Basics</title>


<!-- highlightjs -->
<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>

<!-- taken from https://github.com/rstudio/rmarkdown/blob/67b7f5fc779e4cfdfd0f021d3d7745b6b6e17149/inst/rmd/h/default.html#L296-L362 -->
<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("section-TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>
<!-- end tabsets -->



</head>

<body>



<div class="pageContent band">
<div class="bandContent page">

<div class="topics">

<div id="section-what-are-k-mers" class="section level2">
<h2>What Are K-mers?</h2>
<p>K-mers are DNA sequences of length k. The can be generated by moving along a mother- or template DNA in steps of 1. For k = 1 - the shortest - this equals the set of nucleotides A, C, G, and T. Other well-established k-mers include dimers and trimers. Here we are interested in longer k-mers typically in the range from k &gt; 3 to k-mers of lenght &lt; 100 or also occasionally longer.</p>
<div id="section-what-are-they-good-for" class="section level3">
<h3>What are they good for?</h3>
<p>K-mers have found a variety of applications. Particularly interesting is their use as fingerprints at different levels of resolution. In metagenomics, for example, the mixed sample of DNA from a host of organsims is typically annotated with species labels by screening against a database of k-mer signatures.</p>
</div>
</div>
<div id="section-main-challenges" class="section level2">
<h2>Main challenges</h2>
<p>A major challenge is to decide on a minimal k-mer size that still is capable to support the envisioned application. The longer the k-mers become the more combinations are possible and so an index in a database will become very big or too big. Along with this, the time to create and consult the database will increase. While this can be remedied to a certain extent with parallel computing, it is still desirable and cheaper, to keep the k-mer size small.</p>
<p>Due to the many still unknown factors in genome biology it is yet not possible to calculate exactly how many distinct k-mers to expect for a certain k and a given species. In general we expect, with increasing k to have more unique k-mers. We also expect, that within a species the unqiue k-mers are limited.</p>
<p>In principle, it is possible to heuristically estimate the total number of k-mers, the distinct, and the unique k-mers by systematically scanning a genome for a certain k range. Experience has shown that it is convenient to only consider kmers of of uneven numbers.</p>
<p>Here, the example of the virus phix174 (the first sequenced organism) shows the principle following a blogpost here(x). Another wording is, to evaluate the genomic response to changes in k.</p>
</div>
<div id="section-genomic-response-of-phi-x-174" class="section level2">
<h2>Genomic Response of Phi x 174</h2>
<p>This package contains the full sequende of PhiX174 from NCBI.</p>
<pre class="r"><code>library(kmerize)
library(ggplot2)
library(magrittr)
</code></pre>
<pre class="r"><code>  # Getting a genome
  fa &lt;-  system.file(&quot;testdata/phix174.fasta&quot;, 
                     package = &quot;kmerize&quot;)
                                          
  phix_length &lt;- kmr_length_fasta(fa)                   

  phix_length
#&gt; [1] 5386
  
  # Setting an output file in a temporary directory since this is for demo
  out_file &lt;- file.path(tempdir(), &quot;phix174&quot;)
  
  # here we create a sequence of k in steps of 2

  k &lt;- as.integer(seq(3, 21, 2))
  
  phix_response &lt;- kmr_response(fa, k)
#&gt; C:/Users/Reinhard/AppData/Roaming/WUR/kmerize/tools/kmc3/kmer_counter.exe -k3 -m12 -fm -ci0 -cx1000000000 -cs255 -n255 -t4 D:/apps/R/R-3.6.1/library/kmerize/testdata/phix174.fasta C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/phix174 C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/_kmer
#&gt; 5
#&gt; C:/Users/Reinhard/AppData/Roaming/WUR/kmerize/tools/kmc3/kmer_counter.exe -k5 -m12 -fm -ci0 -cx1000000000 -cs255 -n255 -t4 D:/apps/R/R-3.6.1/library/kmerize/testdata/phix174.fasta C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/phix174 C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/_kmer
#&gt; 7
#&gt; C:/Users/Reinhard/AppData/Roaming/WUR/kmerize/tools/kmc3/kmer_counter.exe -k7 -m12 -fm -ci0 -cx1000000000 -cs255 -n255 -t4 D:/apps/R/R-3.6.1/library/kmerize/testdata/phix174.fasta C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/phix174 C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/_kmer
#&gt; 9
#&gt; C:/Users/Reinhard/AppData/Roaming/WUR/kmerize/tools/kmc3/kmer_counter.exe -k9 -m12 -fm -ci0 -cx1000000000 -cs255 -n255 -t4 D:/apps/R/R-3.6.1/library/kmerize/testdata/phix174.fasta C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/phix174 C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/_kmer
#&gt; 11
#&gt; C:/Users/Reinhard/AppData/Roaming/WUR/kmerize/tools/kmc3/kmer_counter.exe -k11 -m12 -fm -ci0 -cx1000000000 -cs255 -n255 -t4 D:/apps/R/R-3.6.1/library/kmerize/testdata/phix174.fasta C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/phix174 C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/_kmer
#&gt; 13
#&gt; C:/Users/Reinhard/AppData/Roaming/WUR/kmerize/tools/kmc3/kmer_counter.exe -k13 -m12 -fm -ci0 -cx1000000000 -cs255 -n255 -t4 D:/apps/R/R-3.6.1/library/kmerize/testdata/phix174.fasta C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/phix174 C:/Users/Reinhard/AppData/Local/Temp/RtmpkN1yJS/_kmer</code></pre>
</div>
<div id="section-genomic-response-table" class="section level2">
<h2>Genomic Response Table</h2>
<p>The convenience method for the genomic response returns a summary table. This includes also the theoretical maximum number of k-mers (the k-mer universe) for each value of k.</p>
<p>The resulting table shows only values up to 13. This is because currently, it stops once the number of unique k-mers equals (exactly more than .9999) the distinct k-mers. This means, at and after this k value all k-mers have become unique.</p>
<pre class="r"><code>
knitr::kable(phix_response)</code></pre>
<table>
<thead>
<tr class="header">
<th align="right">k</th>
<th align="right">unique</th>
<th align="right">distinct</th>
<th align="right">total</th>
<th align="right">universe</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">3</td>
<td align="right">0</td>
<td align="right">32</td>
<td align="right">5328</td>
<td align="right">32</td>
</tr>
<tr class="even">
<td align="right">5</td>
<td align="right">10</td>
<td align="right">506</td>
<td align="right">5382</td>
<td align="right">512</td>
</tr>
<tr class="odd">
<td align="right">7</td>
<td align="right">2295</td>
<td align="right">3547</td>
<td align="right">5380</td>
<td align="right">8192</td>
</tr>
<tr class="even">
<td align="right">9</td>
<td align="right">4972</td>
<td align="right">5170</td>
<td align="right">5378</td>
<td align="right">131072</td>
</tr>
<tr class="odd">
<td align="right">11</td>
<td align="right">5346</td>
<td align="right">5361</td>
<td align="right">5376</td>
<td align="right">2097152</td>
</tr>
<tr class="even">
<td align="right">13</td>
<td align="right">5374</td>
<td align="right">5374</td>
<td align="right">5374</td>
<td align="right">33554432</td>
</tr>
</tbody>
</table>
</div>
<div id="section-genomic-response-graph" class="section level2">
<h2>Genomic Response Graph</h2>
<p>Plotting this table shows the following response curve:</p>
<pre class="r"><code>

 p &lt;- kmr_plot_response(phix_response, max_y = phix_length)

  p
#&gt; Warning: Removed 3 rows containing missing values (geom_path).</code></pre>
<p><img src="Tutorial_Basics_files/figure-html/grplot-1.png" width="672" /></p>
<p>Now the curves for the unique and distinct k-mers follow an S curve or power law curve with diminishing response in variability (distinct / unique) as k increases. At around 95% there seems to be a distinct land mark.</p>
<p>It could be an example of the 95/5 rule going back to Deming: He argued that 95% of an organization’s performance depends on the quality of the system, and that only 5% is attributable to the actions of discrete individuals. In analogy, one could expect that most of an organisms makeup is needed for basic house keeping and maybe 5% are interesting ‘diversity’.</p>
<pre><code>#&gt; [1] 5112.9</code></pre>
<p>So, this cutoff value can be added to the graph as follows:</p>
<pre class="r"><code>  p &lt;- p +
      geom_hline(yintercept = k_lim, linetype = &quot;dashed&quot;, color = &quot;darkgreen&quot;, lwd = .6)
 p 
#&gt; Warning: Removed 3 rows containing missing values (geom_path).</code></pre>
<p><img src="Tutorial_Basics_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<pre class="r"><code>  p &lt;-  kmr_plot_response(phix_response, max_y = phix_length, limit = lim)
 p
#&gt; Warning: Removed 3 rows containing missing values (geom_path).</code></pre>
<p><img src="Tutorial_Basics_files/figure-html/unnamed-chunk-5-1.png" width="672" /> And the corresponding k value can be extracted with the following function:</p>
</div>
<div id="section-convencience-graph-options" class="section level2">
<h2>Convencience Graph Options</h2>
<p>The plot function includes options to automatically show a k cutoff corresponding to a certain limit and also to set a maximum limit for y axis.</p>
<p>```{ plot3r}</p>
<p>p &lt;- kmr_plot_response(phix_response, limit = lim, max_y = phix_length) p ```</p>

<script type="application/shiny-prerendered" data-context="server-start">
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

library(kmerize)
library(ggplot2)
library(magrittr)

</script>
 
<script type="application/shiny-prerendered" data-context="data">
  # Getting a genome
  fa <-  system.file("testdata/phix174.fasta", 
                     package = "kmerize")
                                          
  phix_length <- kmr_length_fasta(fa)                   

  phix_length
  
  # Setting an output file in a temporary directory since this is for demo
  out_file <- file.path(tempdir(), "phix174")
  
  # here we create a sequence of k in steps of 2

  k <- as.integer(seq(3, 21, 2))
  
  phix_response <- kmr_response(fa, k)

</script>
 <!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootstrap"]},{"type":"character","attributes":{},"value":["3.3.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/bootstrap"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["viewport"]}},"value":[{"type":"character","attributes":{},"value":["width=device-width, initial-scale=1"]}]},{"type":"character","attributes":{},"value":["js/bootstrap.min.js","shim/html5shiv.min.js","shim/respond.min.js"]},{"type":"character","attributes":{},"value":["css/cerulean.min.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["pagedtable"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pagedtable-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["js/pagedtable.js"]},{"type":"character","attributes":{},"value":["css/pagedtable.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["textmate.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-format"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmarkdown/templates/tutorial/resources"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-format.js"]},{"type":"character","attributes":{},"value":["tutorial-format.css","rstudio-theme.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["navigation"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/navigation-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tabsets.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["default.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.1"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71]}},"value":[{"type":"character","attributes":{},"value":["assertthat","backports","base","BiocGenerics","Biostrings","bitops","colorspace","compiler","crayon","datasets","digest","dplyr","evaluate","farver","fastmap","GenomeInfoDb","GenomeInfoDbData","GenomicRanges","ggplot2","glue","graphics","grDevices","grid","gtable","highr","htmltools","htmlwidgets","httpuv","IRanges","jsonlite","kmerize","knitr","labeling","later","lazyeval","learnr","lifecycle","magrittr","markdown","methods","mime","munsell","parallel","pillar","pkgconfig","promises","purrr","R6","rappdirs","Rcpp","RCurl","rlang","rmarkdown","rprojroot","S4Vectors","scales","shiny","stats","stats4","stringi","stringr","tibble","tidyselect","tools","utils","withr","xfun","xtable","XVector","yaml","zlibbioc"]},{"type":"character","attributes":{},"value":["0.2.1","1.1.5","3.6.1","0.30.0","2.52.0","1.0-6","1.4-1","3.6.1","1.3.4","3.6.1","0.6.23","0.8.4","0.14","2.0.3","1.0.1","1.20.0","1.2.1","1.36.1","3.2.1","1.3.1","3.6.1","3.6.1","3.6.1","0.3.0","0.8","0.4.0","1.5.1","1.5.2","2.18.2","1.6.1","0.1.0","1.28","0.3","1.0.0","0.2.2","0.10.0","0.1.0","1.5","1.1","3.6.1","0.9","0.5.0","3.6.1","1.4.3","2.0.3","1.1.0","0.3.3","2.4.1","0.3.1","1.0.3","1.98-1.1","0.4.4","2.1","1.3-2","0.22.1","1.1.0","1.4.0","3.6.1","3.6.1","1.4.5","1.4.0","2.1.3","1.0.0","3.6.1","3.6.1","2.1.2","0.12","1.8-4","0.24.0","2.2.1","1.30.0"]}]}]}
</script>
<!--/html_preserve-->
</div>

</div> <!-- topics -->

<div class="topicsContainer">
<div class="topicsPositioner">
<div class="band">
<div class="bandContent topicsListContainer">

<!-- begin doc-metadata -->
<div id="doc-metadata">
<h2 class="title toc-ignore" style="display:none;">K-mer Basics</h2>
</div>
<!-- end doc-metadata -->

</div> <!-- bandContent.topicsListContainer -->
</div> <!-- band -->
</div> <!-- topicsPositioner -->
</div> <!-- topicsContainer -->


</div> <!-- bandContent page -->
</div> <!-- pageContent band -->




<script>
// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>


</body>

</html>
