<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tutorial_Basics • kmerize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial_Basics">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">kmerize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Tutorial_Basics.html">Tutorial_Basics</a>
    </li>
    <li>
      <a href="../articles/Tutorial_install.html">Tutorial_install</a>
    </li>
    <li>
      <a href="../articles/Tutorial_potato_1.html">Tutorial_potato_1</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Tutorial_Basics</h1>
            
      
      
      <div class="hidden name"><code>Tutorial_Basics.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(kmerize)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)</span></code></pre></div>
<div id="what-are-k-mers" class="section level2">
<h2 class="hasAnchor">
<a href="#what-are-k-mers" class="anchor"></a>What Are K-mers?</h2>
<p>K-mers are DNA sequences of length k. The can be generated by moving along a mother- or template DNA in steps of 1. For k = 1 - the shortest - this equals the set of nucleotides A, C, G, and T. Other well-established k-mers include dimers and trimers. Here we are interested in longer k-mers typically in the range from k &gt; 3 to k-mers of lenght &lt; 100 or also occasionally longer.</p>
<div id="what-are-they-good-for" class="section level3">
<h3 class="hasAnchor">
<a href="#what-are-they-good-for" class="anchor"></a>What are they good for?</h3>
<p>K-mers have found a variety of applications. Particularly interesting is their use as fingerprints at different levels of resolution. In metagenomics, for example, the mixed sample of DNA from a host of organsims is typically annotated with species labels by screening against a database of k-mer signatures.</p>
</div>
</div>
<div id="main-challenges" class="section level2">
<h2 class="hasAnchor">
<a href="#main-challenges" class="anchor"></a>Main challenges</h2>
<p>A major challenge is to decide on a minimal k-mer size that still is capable to support the envisioned application. The longer the k-mers become the more combinations are possible and so an index in a database will become very big or too big. Along with this, the time to create and consult the database will increase. While this can be remedied to a certain extent with parallel computing, it is still desirable and cheaper, to keep the k-mer size small.</p>
<p>Due to the many still unknown factors in genome biology it is yet not possible to calculate exactly how many distinct k-mers to expect for a certain k and a given species. In general we expect, with increasing k to have more unique k-mers. We also expect, that within a species the unqiue k-mers are limited.</p>
<p>In principle, it is possible to heuristically estimate the total number of k-mers, the distinct, and the unique k-mers by systematically scanning a genome for a certain k range. Experience has shown that it is convenient to only consider kmers of of uneven numbers.</p>
<p>Here, the example of the virus phix174 (the first sequenced organism) shows the principle following a blogpost here(x). Another wording is, to evaluate the genomic response to changes in k.</p>
<div id="genomic-response-of-phi-x-174" class="section level3">
<h3 class="hasAnchor">
<a href="#genomic-response-of-phi-x-174" class="anchor"></a>Genomic Response of Phi x 174</h3>
<p>This package contains the full sequende of PhiX174 from NCBI.</p>
<p>These k values are: 5, 7, 9, 11, 13, 15, 17, 19, 21.</p>
<p>The convenience method for the genomic response returns a summary table. This includes also the theoretical maximum number of k-mers (the k-mer universe) for each value of k.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>phix_response &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmr_response.html">kmr_response</a></span>(fa, k)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="co">#&gt; 7</span></span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co">#&gt; 9</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co">#&gt; 11</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="co">#&gt; 13</span></span></code></pre></div>
<p>The resulting table shows only values up to 13. This is because currently, it stops once the number of unique k-mers equals (exactly more than .9999) the distinct k-mers. This means, at and after this k value all k-mers have become unique.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>knitr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(phix_response)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">k</th>
<th align="right">unique</th>
<th align="right">distinct</th>
<th align="right">total</th>
<th align="right">universe</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">5</td>
<td align="right">10</td>
<td align="right">506</td>
<td align="right">5382</td>
<td align="right">512</td>
</tr>
<tr class="even">
<td align="right">7</td>
<td align="right">2295</td>
<td align="right">3547</td>
<td align="right">5380</td>
<td align="right">8192</td>
</tr>
<tr class="odd">
<td align="right">9</td>
<td align="right">4972</td>
<td align="right">5170</td>
<td align="right">5378</td>
<td align="right">131072</td>
</tr>
<tr class="even">
<td align="right">11</td>
<td align="right">5346</td>
<td align="right">5361</td>
<td align="right">5376</td>
<td align="right">2097152</td>
</tr>
<tr class="odd">
<td align="right">13</td>
<td align="right">5374</td>
<td align="right">5374</td>
<td align="right">5374</td>
<td align="right">33554432</td>
</tr>
<tr class="even">
<td align="right">Plott</td>
<td align="right">ing this</td>
<td align="right">table shows</td>
<td align="right">the fol</td>
<td align="right">lowing response curve:</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmr_plot_response.html">kmr_plot_response</a></span>(phix_response)</span>
<span id="cb4-2"><a href="#cb4-2"></a>p</span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">#&gt; geom_path: Each group consists of only one observation. Do you need to adjust</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">#&gt; the group aesthetic?</span></span></code></pre></div>
<p><img src="Tutorial_Basics_files/figure-html/unnamed-chunk-5-1.png" width="672"> ### Chosing a cutoff value</p>
<p>Now the curves for the unique and distinct k-mers follow an S curve or power law curve with diminishing response in variability (distinct / unique) as k increases. At around 95% there seems to be a distinct land mark.</p>
<p>It could be an example of the 95/5 rule going back to Deming: He argued that 95% of an organization’s performance depends on the quality of the system, and that only 5% is attributable to the actions of discrete individuals. In analogy, one could expect that most of an organisms makeup is needed for basic house keeping and maybe 5% are interesting ‘diversity’.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a></span>
<span id="cb5-2"><a href="#cb5-2"></a>lim &lt;-<span class="st"> </span><span class="fl">.95</span></span>
<span id="cb5-3"><a href="#cb5-3"></a>kc_<span class="dv">9</span> &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(phix_response<span class="op">$</span>total) <span class="op">*</span><span class="st"> </span>lim</span>
<span id="cb5-4"><a href="#cb5-4"></a></span>
<span id="cb5-5"><a href="#cb5-5"></a>kc_<span class="dv">9</span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="co">#&gt; [1] 5112.9</span></span></code></pre></div>
<p>So, this cutoff value can be added to the graph as follows:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a> p &lt;-<span class="st"> </span>p <span class="op">+</span></span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="st">      </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> kc_<span class="dv">9</span>, <span class="dt">linetype =</span> <span class="st">"dashed"</span>, <span class="dt">color =</span> <span class="st">"darkgreen"</span>, <span class="dt">lwd =</span> <span class="fl">.6</span>)</span>
<span id="cb6-3"><a href="#cb6-3"></a> p</span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">#&gt; geom_path: Each group consists of only one observation. Do you need to adjust</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt; the group aesthetic?</span></span></code></pre></div>
<p><img src="Tutorial_Basics_files/figure-html/unnamed-chunk-7-1.png" width="672"> And the corresponding k value can be extracted as follows adding a correction of the step size value (2).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>k_<span class="dv">9</span> &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(phix_response[phix_response<span class="op">$</span>distinct <span class="op">&lt;</span><span class="st"> </span>kc_<span class="dv">9</span>, <span class="st">"k"</span>]) <span class="op">+</span><span class="st"> </span>step</span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a>k_<span class="dv">9</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">#&gt; [1] 9</span></span></code></pre></div>
<p>This gives a k-value that captures the k value in the genomic response of phix174 after which diversity decreases rapidly.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>phix_response[phix_response<span class="op">$</span>k <span class="op">&gt;=</span><span class="st"> </span>(k_<span class="dv">9</span> <span class="op">-</span><span class="st"> </span><span class="dv">2</span>), <span class="st">"unique"</span>] <span class="op">/</span><span class="st"> </span>phix_response[phix_response<span class="op">$</span>k <span class="op">&gt;=</span><span class="st"> </span>(k_<span class="dv">9</span> <span class="op">-</span><span class="st"> </span><span class="dv">2</span>), <span class="st">"distinct"</span>]</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co">#&gt; [1] 0.6470257 0.9617021 0.9972020 1.0000000</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>k_lim =<span class="st"> </span><span class="dv">9</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>yk =<span class="st"> </span><span class="dv">2000</span></span>
<span id="cb9-3"><a href="#cb9-3"></a>p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmr_plot_response.html">kmr_plot_response</a></span>(phix_response, <span class="dt">limit =</span> lim)</span>
<span id="cb9-4"><a href="#cb9-4"></a>p </span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="co">#&gt; geom_path: Each group consists of only one observation. Do you need to adjust</span></span>
<span id="cb9-6"><a href="#cb9-6"></a><span class="co">#&gt; the group aesthetic?</span></span></code></pre></div>
<p><img src="Tutorial_Basics_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#what-are-k-mers">What Are K-mers?</a></li>
      <li><a href="#main-challenges">Main challenges</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Reinhard Simon.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
